src/ledger/user.js

```diff
@@ -0,0 +1,105 @@
+// @flow
+
+/**
+ * This module has a core data type identifying SourceCred users.
+ *
+ * The scope for this data type is to model:
+ * - a unique identifier for each user
+ * - a unique (renameable) username they choose
+ * - the address of every node they correspond to in the graph
+ *
+ * Unlike most other state in SourceCred, the User state is
+ * nondeterministically generated by SourceCred itself, and then persisted
+ * long-term within the instance.
+ *
+ * This is in contrast to Graph data, which usually comes from an external
+ * source, and is not persisted long-term, but instead is re-generated when
+ * needed.
+ *
+ * In particular, this kernel of user data is stored within the core ledger,
+ * since it's necessary to track consistently when tracking Grain distribution.
+ * This type should not grow to include all the data that the UI will
+ * eventually want to show; that should be kept in a different data store which
+ * isn't being used as a transaction ledger.
+ *
+ */
+import {
+  type Uuid,
+  parser as uuidParser,
+  random as randomUuid,
+} from "../util/uuid";
+import * as C from "../util/combo";
+import {
+  type NodeAddressT,
+  NodeAddress,
+  type Node as GraphNode,
+} from "../core/graph";
+
+/**
+ * We validate usernames using GitHub-esque rules.
+ */
+export opaque type Username: string = string;
+const USERNAME_PATTERN = /^@?([A-Za-z0-9-_]+)$/;
+
+export type UserId = Uuid;
+export type User = {|
+  // UUID, assigned when the user is created.
+  +id: UserId,
+  +name: Username,
+  // Every other node in the graph that this user corresponds to.
+  // Does not include the user's "own" address, i.e. the result
+  // of calling (userAddress(user.id)).
+  +aliases: $ReadOnlyArray<NodeAddressT>,
```

Comment:
Late comment, but why not implement this as a Set?
It seems order isn't important, and duplicates don't make sense.

Reply:
Good question. I thought about this, but if using a set, we'd can no longer serialize users via `stringify(user)`, we'd have to have the `User` type and `UserSerialized` with converters between. Didn't seem like the Set added enough to justify the added complexity.

Reply:
True, one way of this conversion can be handled by the parser, but `stringify` wouldn't unless it implemented a `toJSON` method. Fortunately those are easier to validate as we have the JsonObject we can set as return type now.

I'll leave it as a suggestion, but I suspect we'll end up with a serialization layer either way. Event types will probably include a fair bit of parsing logic. Serializing sets by comparison don't seem hard.